File: src/main/java/io/github/saneea/feature/binary/Hash.java
package io.github.saneea.feature.binary;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.PrintStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Option;

import io.github.saneea.Feature;
import io.github.saneea.FeatureContext;

public class Hash implements Feature, Feature.CLI, Feature.In.Bin.Stream, Feature.Out.Text.PrintStream {

	private static int BUFFER_SIZE = 4096;

	private CommandLine commandLine;

	private InputStream in;
	private PrintStream out;

	@Override
	public String getShortDescription() {
		return "calc hash (md5, sha-* etc.)";
	}

	private static void execute(InputStream input, PrintStream out, String alg)
			throws NoSuchAlgorithmException, IOException {

		MessageDigest md = MessageDigest.getInstance(alg);

		byte[] buf = new byte[BUFFER_SIZE];

		int len;
		while ((len = input.read(buf)) != -1) {
			md.update(buf, 0, len);
		}

		ToHex.run(new ByteArrayInputStream(md.digest()), out);
	}

	@Override
	public void run(FeatureContext context) throws NoSuchAlgorithmException, IOException {
		String alg = commandLine.getOptionValue(Params.ALGORITHM);
		execute(in, out, alg);
	}

	@Override
	public Option[] getOptions() {
		return Params.createOptions();
	}

	@Override
	public void setCommandLine(CommandLine commandLine) {
		this.commandLine = commandLine;
	}

	public static class Params {

		public static String ALGORITHM = "algorithm";

		private static Option[] createOptions() {

			Option[] options = { //
					Option//
							.builder("a")//
							.longOpt(ALGORITHM)//
							.hasArg(true)//
							.argName("algorithm name")//
							.required(true)//
							.desc("hash algorithm (e.g. md5)")//
							.build() };

			return options;
		}
	}

	@Override
	public void setOut(PrintStream out) {
		this.out = out;
	}

	@Override
	public void setIn(InputStream in) {
		this.in = in;
	}
}

File: src/main/kotlin/io/github/saneea/dvh/feature/binary/Hash.kt
package io.github.saneea.dvh.feature.binary

import io.github.saneea.dvh.Feature
import io.github.saneea.dvh.Feature.CLI
import io.github.saneea.dvh.Feature.Meta
import io.github.saneea.dvh.feature.binary.hex.ToHex
import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.Option
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.io.PrintStream
import java.security.MessageDigest

class Hash :
    Feature,
    CLI,
    CLI.Options,
    Feature.In.Bin.Stream,
    Feature.Out.Text.PrintStream {

    override lateinit var commandLine: CommandLine
    override lateinit var inBinStream: InputStream
    override lateinit var outTextPrintStream: PrintStream

    override val meta = Meta("calc hash (md5, sha-* etc.)")

    override fun run() {
        val alg = commandLine.getOptionValue(Params.ALGORITHM)
        execute(inBinStream, outTextPrintStream, alg)
    }

    override val options = Params.createOptions()

    object Params {
        const val ALGORITHM = "algorithm"
        fun createOptions() = listOf(
            Option
                .builder("a")
                .longOpt(ALGORITHM)
                .hasArg(true)
                .argName("algorithm name")
                .required(true)
                .desc("hash algorithm (e.g. md5)")
                .build()
        )
    }

    companion object {
        private const val BUFFER_SIZE = 4096

        private fun execute(input: InputStream, out: PrintStream, alg: String) {
            val md = MessageDigest.getInstance(alg)
            val buf = ByteArray(BUFFER_SIZE)
            var len: Int
            while (input.read(buf).also { len = it } != -1) {
                md.update(buf, 0, len)
            }
            ToHex.run(ByteArrayInputStream(md.digest()), out)
        }
    }
}